### PR TITLE
Check incoming relay requests events against the chain

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs #1690

If there are some problems with ethereum node synchronization it may happen that multiple blocks arrive at the same time. It may happen that those blocks contain multiple relay entry request events. Currently, each event will be processed by the client. This is wrong because only the latest event is valid from the chain's perspective and only one request can be processed at the same time. 

Here we introduce a condition that checks event block number against the current entry start block from the chain. This way we handle only the latest event and ignore the ones which belong to the past from the chain's viewpoint.

### CODE REVIEW

File: pkg/chain/gen/abi/KeepRandomBeaconOperator.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ var (
...
...e\":\"uint256[]\"}],\"name\":\"submitDkgResult\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"internalType\":\"bytes32\",\"name\":\"ticket\",\"type\":\"bytes32\"}],\"name\":\"submitTicket\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"submittedTickets\",\"outputs\":[{\"internalType\":\"uint64[]\",\"name\":\"\",\"type\":\"uint64[]\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":true,\"inputs\":[],\"name\":\"ticketSubmissionTimeout\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"\",\"type\":\"uint256\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"internalType\":\"address\",\"name\":\"operator\",\"type\":\"address\"},{\"internalType\":\"uint256\",\"name\":\"groupIndex\",\"type\":\"uint256\"},{\"internalType\":\"uint256[]\",\"name\":\"groupMemberIndices\",\"type\":\"uint256[]\"}],\"name\":\"withdrawGroupMemberRewards\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"}]"
```
Comment (pdyraga) [inline pkg/chain/gen/abi/KeepRandomBeaconOperator.go:31]: We will have a conflict with https://github.com/keep-network/keep-core/pull/1708 here and let's use the version from #1708 when resolving. It has all the required functions as well as the fix for `from` caller address.

File: pkg/beacon/relay/chain/chain.go [hunk 2]
```diff
@@ -38,6 +40,9 @@ type RelayEntryInterface interface {
 	// supposed to submit a relay entry, did not deliver it within a specified
 	// time frame (relayEntryTimeout) counted in blocks.
 	ReportRelayEntryTimeout() error
+	// CurrentEntryStartBlock returns the starting block of the
+	// current relay entry.
+	CurrentEntryStartBlock() (*big.Int, error)
```
Comment (pdyraga) [inline pkg/beacon/relay/chain/chain.go:45]: I asked @dimpar to rename it to `CurrentRequestStartBlock` in #1698. I missed to rename it in the contract earlier and we are having a mess in the naming. Everything but this one is `currentRequest*`.

File: pkg/beacon/beacon.go [hunk 3]
```diff
@@ -74,12 +74,33 @@ func Initialize(
 	}
 
 	relayChain.OnRelayEntryRequested(func(request *event.Request) {
-		previousEntry := hex.EncodeToString(request.PreviousEntry[:])
+		currentEntryStartBlock, err := relayChain.CurrentEntryStartBlock()
+		if err != nil {
+			logger.Warningf("could not check current entry start block")
```
Comment (pdyraga) [inline pkg/beacon/beacon.go:79]: `*current request start block` please. We need to rename this function. It's my fault, sorry.

File: pkg/beacon/beacon.go [hunk 4]
```diff
@@ -74,12 +74,33 @@ func Initialize(
...
+		// the current entry start block from the chain. Most likely, there
+		// are some problems with chain synchronization and multiple events
+		// arrived at the same time but only the last should be handled.
+		if request.BlockNumber != currentEntryStartBlock.Uint64() {
+			logger.Warningf(
+				"ignoring relay entry request with block number [%v]; "+
+					"current entry start block stored on-chain is [%v]",
+				request.BlockNumber,
+				currentEntryStartBlock.Uint64(),
+			)
```
Comment (pdyraga) [inline pkg/beacon/beacon.go:93]: Shouldn't we `return` here?

File: pkg/chain/gen/abi/KeepRandomBeaconOperator.go (hunk 1 above)
Comment (lukasz-zimnoch) [reply pkg/chain/gen/abi/KeepRandomBeaconOperator.go:31]: Ok, I'll wait for #1708

File: pkg/beacon/beacon.go (hunk 4 above)
Comment (lukasz-zimnoch) [reply pkg/beacon/beacon.go:93]: Yeah, for sure. Thanks for catching this!

File: pkg/chain/gen/abi/KeepRandomBeaconOperator.go (hunk 1 above)
Comment (pdyraga) [reply pkg/chain/gen/abi/KeepRandomBeaconOperator.go:31]: Even better, please take a look at #1708 and review this code, I think we can merge it pretty quickly.

**Pull request merged**